File: HW5/BoardArray1D.cpp
//*************			TURKER		TERCAN		171044032		HW4		CSE241	OOP				 *************//

//Implementation for AbstractBoard
#include "BoardArray1D.h"
#include <iostream>
#include <fstream>

using namespace std;
namespace TurkerTercanHW5
{
	BoardArray1D::BoardArray1D() : AbstractBoard(){ }
	BoardArray1D::BoardArray1D(int newSize, int newWidth) : AbstractBoard(newSize, newWidth)
	{
		int count = 1;
		board = new int[newSize * newWidth];
		for(int i = 0; i < newSize * newWidth; i++)
		{
			board[i] = count++;
			if( i == (newSize * newWidth) - 1)
				board[i] = -1;
		}
		setLocationX(newSize - 1);
		setLocationY(newWidth - 1);
		randomMoves();
	}
	
	BoardArray1D::BoardArray1D(const AbstractBoard& object) : AbstractBoard(object)
	{
		setSize(object.getSize(), object.getWidth());		
		setLastMove(object.lastMove());
		setNumberOfMoves(object.numberOfMoves());
		setSolved(object.isSolved());
		setLocationX(object.getLocationX());
		setLocationY(object.getLocationY());
		board = new int[getSize() * getWidth()];
		int count = 0;
		for(int i = 0; i < getSize(); i++)
		{
			for(int j = 0; j < getWidth(); j++)
			{
				board[count++] = object(i,j);
			}
		}
	}
	
	BoardArray1D& BoardArray1D::operator =(const AbstractBoard& other)
	{
		free(board);
		setSize(other.getSize(), other.getWidth());		
		setLastMove(other.lastMove());
		setNumberOfMoves(other.numberOfMoves());
		setSolved(other.isSolved());
		setLocationX(other.getLocationX());
		setLocationY(other.getLocationY());
		board = new int[getSize() * getWidth()];
		int count = 0;
		for(int i = 0; i < getSize(); i++)
		{
			for(int j = 0; j < getWidth(); j++)
			{
				board[count++] = other(i,j);
			}
		}
		return *this;
	}
	
	BoardArray1D::~BoardArray1D()
	{
		free(board);
	}
	
	void BoardArray1D::readFromFile(istream& infile)
	{
		free(board);
		setSize(0);
		setWidth(0);
		int temp;
		int i = 0;
		int y = 0;
		const int lineCapacity = 30; //there can be 30 char values in the file
		char line[lineCapacity];
		
		while(!infile.eof())
		{
			infile.getline(line, lineCapacity);
			setSize(getSize() + 1);
		}
		for(int i = 0; i < lineCapacity; i++)
		{
			if(line[i] == '\n' || line[i] == '\0')
			{
				setWidth( getWidth() + 1);
				break;
			}
			else
				setWidth( getWidth() + 1);
		}
		setWidth( getWidth() / 3);
		infile.clear();
		infile.seekg(0, ios::beg);
		
		board = new int[getSize() * getWidth()];
		while(!infile.eof())
		{
			infile.getline(line, lineCapacity);
			for(int j = 0; j < lineCapacity; j+=3)
			{
				temp = (((int)line[j] - 48) * 10) + (int)line[j+1] - 48;
				if(temp == 550)//int value of bb
				{	
					temp = -1;
					setLocationX(i);
					setLocationY(y);
				}
				board[i * getWidth() + y++] = temp;
			}
			y = 0;
			i++;
		}
	}
	
	void BoardArray1D::move(char choice)
	{
		switch(choice){
			case 'L': moveLeft();
			break;
			case 'R': moveRight();
			break;
			case 'U': moveUp();
			break;
			case 'D': moveDown();
			break;
			default: cout << "Wrong decision" << endl;
		}
	}
	
	void BoardArray1D::swapBlank(int x, int y)
	{
		board[ getLocationX() * getWidth() + getLocationY()] = board[ x * getWidth() + y];
		board[ x * getWidth() + y] = -1;
		setLocationX(x);
		setLocationY(y);
	}
	
	void BoardArray1D::moveLeft()
	{
		if(getLocationY() != 0 && board[(getLocationX() * getWidth()) + getLocationY() - 1] != 0){
			swapBlank(getLocationX(), getLocationY() - 1);
			setNumberOfMoves(numberOfMoves() + 1);
			setLastMove('L');
			setSolved(check());
		}
		else if( getLocationY() == 0 )
			cout << "Unavailable to move: Left side is empty" << endl;
		else if( board[(getLocationX() * getWidth()) + getLocationY() - 1] == 0 )
			cout << "Unavailable to move: Left side is a wall" << endl;
	}
	
	void BoardArray1D::moveRight()
	{
		if(getLocationY() != getWidth() - 1 && board[(getLocationX() * getWidth()) + getLocationY() + 1] != 0){
			swapBlank(getLocationX(), getLocationY() + 1);
			setNumberOfMoves(numberOfMoves() + 1);
			setLastMove('R');
			setSolved(check());
		}
		else if( getLocationY() == getWidth() - 1 )
			cout << "Unavailable to move: Right side is empty" << endl;
		else if( board[(getLocationX() * getWidth()) + getLocationY() + 1] == 0 )
			cout << "Unavailable to move: Right side is a wall" << endl;
	}
	
	void BoardArray1D::moveUp()
	{
		if(getLocationX() != 0 && board[((getLocationX() - 1) * getWidth()) + getLocationY() ] != 0){
			swapBlank(getLocationX() - 1, getLocationY());
			setNumberOfMoves(numberOfMoves() + 1);
			setLastMove('U');
			setSolved(check());
		}
		else if( getLocationX() == 0 )
			cout << "Unavailable to move: Up side is empty" << endl;
		else if( board[((getLocationX() - 1) * getWidth()) + getLocationY() ] == 0 )
			cout << "Unavailable to move: Up side is a wall" << endl;
	}
	
	void BoardArray1D::moveDown()
	{
		if(getLocationX() != getSize() - 1 && board[((getLocationX() + 1) * getWidth()) + getLocationY() ] != 0){
			swapBlank(getLocationX() + 1, getLocationY());
			setLastMove('D');
			setNumberOfMoves(numberOfMoves() + 1);
			setSolved(check());
		}
		else if( getLocationX() == getSize() - 1 )
			cout << "Unavailable to move: Down side is empty" << endl;
		else if( board[((getLocationX() + 1) * getWidth()) + getLocationY() ] == 0 )
			cout << "Unavailable to move: Down side is a wall" << endl;
	}
	
	int* BoardArray1D::getBoard() const
	{
		return board;
	}
	
	int BoardArray1D::operator ()(int x, int y) const
	{
		return board[ x * getWidth() + y];
	}
	
	int BoardArray1D::getTile(int x, int y) const
	{
		return board[ x * getWidth() + y];
	}
	
	void BoardArray1D::setTile(int tile, int x, int y)
	{
		board[x * getWidth() + y] = tile;
	}
	
	void BoardArray1D::randomMoves()
	{
		srand(time(NULL));
		int n;
		for(int i = 0; i < getSize() * getWidth(); i++){
			n = rand() % 4;
			if( n == 0 && lastMove() != 'D' && getLocationX() != 0)
				move('U');
			else if( n == 1 && lastMove() != 'R' && getLocationY() != 0)
				move('L');
			else if( n == 2 && lastMove() != 'L' && getLocationY() != getWidth() -1)
				move('R');
			else if( n == 3 && lastMove() != 'U' && getLocationX() != getSize() - 1)
				move('D');
		}
	}
	
	bool BoardArray1D::check() const
	{
		int count = 1;
		for(int i = 0; i < getSize() * getWidth(); i++)
		{
			if( i == getSize() * getWidth()- 1  && board[i] == -1)
				return 1;
			if( board[i] != count++)
				return 0;
		}
		return 1;
	}
}

File: HW7/src/hw07/BoardArray1D.java
package hw07;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Random;
import java.util.Scanner;

public class BoardArray1D extends AbstractBoard{
	
	/**
	 * One dimensional array
	 */
	private int[] board;

	/*
	 * When i tried to use two different scanners for write and read functions, i'd received NoSuchElementException
	 * to avoid that i used one scanner for both.
	 */
	private Scanner terminal = new Scanner(System.in);
	
	/**
	 * Swaps BB to that tile
	 * @param x is tile's x coordinate
	 * @param y is tile's y coordinate
	 */
	private void swapBlank( int x, int y ) 
	{
		board[this.getLocationX() * this.getWidth() + this.getLocationY()] = this.cell(x, y);
		board[x * getWidth() + y] = -1;
		this.setLocationX( x );
		this.setLocationY( y );
	}
	
	/**
	 * BB's move down method
	 * It will not allow used to moveDown if there is a wall or empty
	 */
	private void moveDown() 
	{
		if( this.getLocationX() != this.getSize() - 1 && this.cell( this.getLocationX( ) + 1, this.getLocationY( ) ) != 0 )
		{	
			this.swapBlank( this.getLocationX( ) + 1, this.getLocationY( ) );
			this.increaseMoves();
			this.setLast('D');
			this.checkIsFinal();
		}
		else if( this.getLocationX() == this.getSize() - 1 )
			System.out.println("Unavailable to move: Down side is empty");
		else if( this.cell( getLocationX( ) + 1, getLocationY( ) ) == 0 )
			System.out.println("Unavailable to move: Down side is a wall");
	}
	
	/**
	 * BB's move up method
	 * It will not allow used to moveUp if there is a wall or empty
	 */
	private void moveUp() 
	{
		if( this.getLocationX() != 0 && this.cell( this.getLocationX( ) - 1, this.getLocationY( ) ) != 0)
		{
			this.swapBlank(getLocationX() -1, getLocationY());
			this.increaseMoves();
			this.setLast('U');
			this.checkIsFinal();
		}
		else if( this.getLocationX() == 0 )
			System.out.println("Unavailable to move: Up side is empty");
		else if( this.cell(getLocationX() -1, getLocationY()) == 0)
			System.out.println("Unavailable to move: Up side is a wall");
	}
	
	/**
	 * BB's move right method
	 * It will not allow used to moveRight if there is a wall or empty
	 */
	private void moveRight() 
	{
		if( this.getLocationY() != this.getWidth() - 1 && this.cell(getLocationX(), getLocationY() + 1) != 0 )
		{
			this.swapBlank(getLocationX(), getLocationY() + 1);
			this.increaseMoves();
			this.setLast('R');
			this.checkIsFinal();
		}
		else if( this.getLocationY() == this.getWidth() - 1 )
			System.out.println("Unavailable to move: Right side is empty");
		else if( this.cell(getLocationX(), getLocationY() + 1) == 0 )
			System.out.println("Unavailable to move: Right side is a wall");
	}
	
	/**
	 * BB's move left method
	 * It will not allow used to moveLeft if there is a wall or empty
	 */
	private void moveLeft() 
	{
		if( this.getLocationY() != 0 && this.cell(getLocationX(), getLocationY() - 1) != 0 )
		{
			this.swapBlank(getLocationX(), getLocationY() - 1);
			this.increaseMoves();
			this.setLast('L');
			this.checkIsFinal();
		}
		else if( this.getLocationY() == 0 )
			System.out.println("Unavailable to move: Left side is empty");
		else if( this.cell(getLocationX(), getLocationY() - 1) == 0 )
			System.out.println("Unavailable to move: Left side is a wall");
	}
	
	/**
	 * Basic constructor
	 * @param newX size of x coordinates
	 * @param newY size of y coordinates
	 */
	public BoardArray1D( int newX, int newY )
	{
		super( newX, newY );
		this.randomMoves( );
	}
	
	/**
	 * I used this constructor to get clone of the current board
	 * @param clone clone is the current board
	 */
	public BoardArray1D( AbstractBoard clone )
	{
		super( clone );
	}
	
	/**
	 * it makes 5 random moves
	 */
	public void randomMoves( ) 
	{
		Random rand = new Random( );
		int n;
		for( int i = 0; i < 5 ; i++)
		{
			int x = this.getLocationX();
			int y = this.getLocationY();
			n = rand.nextInt( 4 );
			if( n == 0 & this.lastMove( ) != 'D' && this.getLocationX( ) != 0 )
				this.move( 'U' );
			else if( n == 1 && this.lastMove( ) != 'R' && this.getLocationY() != 0 )
				this.move( 'L' );
			else if( n == 2 && this.lastMove( ) != 'L' && this.getLocationY() != this.getWidth() - 1 )
				this.move( 'R' );
			else if( n == 3 && this.lastMove( ) != 'U' && this.getLocationX() != this.getSize() - 1 )
				this.move( 'D' );
			if( x != this.getLocationX() || y != this.getLocationY() )
				this.setMoves( this.numberOfMoves() - 1);
		}
	}
	
	/**
	 * Initialize one dimensional board
	 */
	public void createBoard( )
	{
		board = new int[ this.getSize( ) * this.getWidth( ) ];
		this.setLocationX( this.getSize( ) - 1 );
		this.setLocationY( this.getWidth( ) - 1 );
	}
	
	/**
	 * This method finds and returns integer in that positon
	 * @param x is x coordinate
	 * @param y is y coordinate
	 * @return integer in that position
	 */
	public int getTile( int x, int y )
	{
		return board[x * getWidth() + y];
	}
	
	@Override
	protected void setTile( int tile, int x, int y )
	{
		board[x * getWidth() + y] = tile;
	}
	
	@Override
	public void move( char direction )
	{
		switch( direction )
		{ 
			case 'L': this.moveLeft( );
			break;
			case 'R': this.moveRight( );
			break;
			case 'U': this.moveUp( );
			break;
			case 'D': this.moveDown( );
			break;
			default: System.out.printf("Wrong Decision!\n");
		}
	}
	
	@Override
	public void writeToFile( )
	{
		System.out.println("Enter your file's \\path\\fileName:");
		String fileName = new String(terminal.nextLine());
		File text = new File(fileName);
		
		try
		{
			if(text.createNewFile())
				System.out.println("File created: " + text.getName());
			else
				System.out.println("File already exist in that location");
			FileWriter writer = new FileWriter(text);
			
			for( int i = 0; i < this.getSize(); i++ )
			{
				for( int j = 0; j < this.getWidth(); j++ )
				{
					if(this.cell(i, j) == -1)
					{
						writer.write("bb ");
					}
					else if( this.cell(i, j) < 10 )
						writer.write("0" + this.cell(i, j) + " ");
					else if( this.cell(i, j) >= 10 )
						writer.write(this.cell(i, j) + " ");
				}
				if( i != this.getSize() - 1)
					writer.write("\n");
			}
			writer.close();
		}
		catch( IOException e )
		{
			System.out.println(e.getMessage() + " writeToFile closes..\n");
		}	
	}
	
	@Override
	public void readFromFile( ) throws FileNotFoundException
	{
		System.out.println("Enter your file's path: ");
		String fileName = new String(terminal.nextLine());
		File text = new File(fileName);
		Scanner reader = new Scanner(text);
		
		int x = 0, y = 0;
		String temp;
		while( reader.hasNext() )
		{
			temp = new String(reader.nextLine());
			x++;
			y = (temp.length() + 1) / 3;
		}
		this.setSize(x, y);
		
		reader.close();
		reader = new Scanner(text);
		x = 0;
		y = 0;
		char c1, c2;
		int val;
		while( reader.hasNext() )
		{
			String read = new String(reader.nextLine());
			y = 0;
			for( int i = 0; i < read.length(); i+=3)
			{
				c1 = read.charAt( i );
				c2 = read.charAt( i + 1 );
				if( c1 == 'b' && c2 == 'b' ) {
					this.setTile( -1, x, y );
					this.setLocationX(x);
					this.setLocationY(y);
				}
				else
				{
					val = Character.getNumericValue(c1) * 10 + Character.getNumericValue(c2);
					this.setTile(val, x, y);
				}
				y++;	
			}
			x++;
		}
		reader.close();
	}
	
	@Override
	public String toString()
	{
		StringBuilder build = new StringBuilder();
		build.append(super.toString());
		
		for( int i = 0; i < this.getSize(); i++ )
		{
			for( int j = 0; j < this.getWidth(); j++ )
			{
				if( this.cell(i, j) == -1 )
					build.append("bb ");
				else if( this.cell(i, j) < 10)
					build.append("0" + this.cell(i, j) + " ");
				else if( this.cell(i, j) >= 10 )
					build.append(this.cell(i, j) + " ");
			}
			build.append("\n");
		}	
		
		return build.toString();
	}
}
